File: Signature/codegen/lib/processSIGburst_onboard/nanmean.cpp
//
// Academic License - for use in teaching, academic research, and meeting
// course requirements at degree granting institutions only.  Not for
// government, commercial, or other organizational use.
//
// nanmean.cpp
//
// Code generation for function 'nanmean'
//

// Include files
#include "nanmean.h"
#include "rt_nonfinite.h"
#include <cmath>

// Function Definitions
double nanmean(double x)
{
  double b_unnamed_idx_0;
  double unnamed_idx_0;
  bool nans;
  // NANMEAN Mean value, ignoring NaNs.
  //    M = NANMEAN(X) returns the sample mean of X, treating NaNs as missing
  //    values.  For vector input, M is the mean value of the non-NaN elements
  //    in X.  For matrix input, M is a row vector containing the mean value of
  //    non-NaN elements in each column.  For N-D arrays, NANMEAN operates
  //    along the first non-singleton dimension.
  //
  //    NANMEAN(X,DIM) takes the mean along dimension DIM of X.
  //
  //    See also MEAN, NANMEDIAN, NANSTD, NANVAR, NANMIN, NANMAX, NANSUM.
  //    Copyright 1993-2004 The MathWorks, Inc.
  //    $Revision: 2.13.4.3 $  $Date: 2004/07/28 04:38:41 $
  //  Find NaNs and set them to zero
  nans = std::isnan(x);
  unnamed_idx_0 = x;
  if (nans) {
    unnamed_idx_0 = 0.0;
  }
  //  Count up non-NaNs.
  b_unnamed_idx_0 = !nans;
  if (nans) {
    b_unnamed_idx_0 = rtNaN;
  }
  //  prevent divideByZero warnings
  //  Sum up non-NaNs, and divide by the number of non-NaNs.
  if (std::isnan(unnamed_idx_0)) {
    unnamed_idx_0 = 0.0;
  }
  return unnamed_idx_0 / b_unnamed_idx_0;
}

// End of code generation (nanmean.cpp)

File: GeneralTools/nanmean.m
function y = nanmean(x, dim)
% NANMEAN    Mean value, same as in matlab 5 but ignoring NaNs
%===================================================================
% NANMEAN   2.1 30/5/97
%
% function y = nanmean(x, dim)
%
% DESCRIPTION:
%    Mean value, same as in matlab 5 but ignoring NaNs.
%    For vectors, NANMEAN(X) is the mean value of the elements in X. For
%    matrices, NANMEAN(X) is a row vector containing the mean value of
%    each column.  For N-D arrays, NANMEAN(X) is the mean value of the
%    elements along the first non-singleton dimension of X.
% 
%    NANMEAN(X,DIM) takes the mean along the dimension DIM of X. 
% 
% INPUT:
%    x    = array of any dimension
%    dims = dimension along which to take the mean
%
% OUTPUT:
%    y    =  a mean of actual data values, i.e., ignoring NaNs.
%
% EXAMPLES:
%           x = [ 1  2  3; 5 NaN 7];
%           y = nanmean(x)
%           y = [3 2 5]
%
%           x = [ 1  NaN  3; 5 NaN 7];
%           y = nanmean(x)
%           y = [3 NaN 5]
%
%           x = [ 1  NaN  3; 5 NaN 7];
%           y = nanmean(x, 2)
%           y = [2; 6]
%
% CALLER:   general purpose
% CALLEE:   none
%
% AUTHOR:   Jim Mansbridge
%==================================================================

% $Id: nanmean.m,v 1.4 1997/06/02 06:00:58 mansbrid Exp $
% 
%--------------------------------------------------------------------

if nargin == 1
  aa = ~isnan(x);
  ff = find(aa == 0);
  if (length(ff) ~= 0) % replace NaNs by zeros
    x(ff) = zeros(size(ff));
  end
  ss = sum(aa);        % find the total number of non-nans in the column
  ff = find(ss == 0);
  if (length(ff) ~= 0) % put NaNs where the mean came from a column of NaNs
    ss(ff) = NaN*zeros(size(ff));
  end
  y = sum(x)./ss;      % calculate the mean
elseif nargin == 2
  aa = ~isnan(x);
  ff = find(aa == 0);
  if (length(ff) ~= 0) % replace NaNs by zeros
    x(ff) = zeros(size(ff));
  end
  ss = sum(aa, dim);   % find the total number of non-nans in the column
  ff = find(ss == 0);
  if (length(ff) ~= 0) % put NaNs where the mean came from a column of NaNs
    ss(ff) = NaN*zeros(size(ff));
  end
  y = sum(x, dim)./ss; % calculate the mean
end
